rust/bridge/node/src/lib.rs

```diff
@@ -6,40 +6,48 @@
 use libsignal_protocol_rust::*;
 use neon::context::Context;
 use neon::prelude::*;
+use std::ops::Deref;
 
-fn borrow_this<'a, V, T, F>(cx: &mut MethodContext<'a, V>, f: F) -> T
-where
-    V: Class,
-    F: for<'b> FnOnce(neon::borrow::Ref<'b, &mut <V as Class>::Internals>) -> T,
-{
-    let this = cx.this();
-    cx.borrow(&this, f)
-}
+struct DefaultFinalize<T>(T);
+
+impl<T> Finalize for DefaultFinalize<T> {}
+
+impl<T> Deref for DefaultFinalize<T> {
+    type Target = T;
 
-declare_types! {
-    pub class JsPrivateKey for PrivateKey {
-        init(_cx) {
-            // FIXME: guard against calling this directly
-            let mut rng = rand::rngs::OsRng;
-            let keypair = KeyPair::generate(&mut rng);
-            Ok(keypair.private_key)
-        }
-
-        method serialize(mut cx) {
-            let bytes = borrow_this(&mut cx, |k| {
-                k.serialize()
-            });
-            // FIXME: check for truncation
-            let mut buffer = cx.buffer(bytes.len() as u32)?;
-            cx.borrow_mut(&mut buffer, |raw_buffer| {
-                raw_buffer.as_mut_slice().copy_from_slice(&bytes);
-            });
-            Ok(buffer.upcast())
-        }
+    fn deref(&self) -> &Self::Target {
+        &self.0
     }
 }
 
-register_module!(mut cx, {
-    cx.export_class::<JsPrivateKey>("PrivateKey")?;
+type DefaultJsBox<T> = JsBox<DefaultFinalize<T>>;
+
+fn boxed<'a, T: Send>(cx: &mut FunctionContext<'a>, value: T) -> Handle<'a, DefaultJsBox<T>> {
+    cx.boxed(DefaultFinalize(value))
+}
+
+#[allow(non_snake_case)]
+fn PrivateKey_generate(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let cx = &mut cx;
+    let mut rng = rand::rngs::OsRng;
+    let keypair = KeyPair::generate(&mut rng);
+    Ok(boxed(cx, keypair.private_key).upcast())
+}
+
+#[allow(non_snake_case)]
+fn PrivateKey_serialize(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let value = cx.argument::<DefaultJsBox<PrivateKey>>(0)?;
+    let bytes = value.serialize();
+    let mut buffer = cx.buffer(bytes.len() as u32)?;
```

Comment:
tl;dr, could you change this to `cx.array_buffer` if possible? Also means updating the TypeScript types.

---

Like much of the JavaScript world, Signal Desktop has historically been inconsistent about which byte sequence type to use. Should we use the [Node-native `Buffer`](https://nodejs.org/api/buffer.html)? The [standard `ArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)? A [userland wrapper](https://github.com/signalapp/Signal-Desktop/blob/a5657473c3b2b36afa1cade62323ed88464be400/components/bytebuffer/dist/ByteBufferAB.js)?

I've asked the rest of the Desktop team what we'd like to standardize on. My preference is `ArrayBuffer` because it is (1) standard across runtimes, unlike `Buffer` (2) not tied to a specific "interpretation" of what the bytes represent, like `Uint8Array`.

For now, I think `ArrayBuffer` is a better choice. If you'd like to wait for our discussion to resolve, I'm fine to leave it as `Buffer`, but we might have to make a small breaking change.

Reply:
I think I had picked Buffer to match what (I think) libtextsecure is currently doing, but I'm happy to switch it now or later.

Reply:
`Buffer` is all good!